python_modules/dagster/dagster/_core/definitions/unresolved_asset_sensor_definition.py

```diff
@@ -0,0 +1,266 @@
+import inspect
+import json
+from collections import defaultdict
+from itertools import chain
+from typing import TYPE_CHECKING, List, Mapping, Optional, Sequence
+
+from dagster._annotations import experimental
+from dagster._core.errors import DagsterInvalidDefinitionError
+
+from .asset_selection import AssetSelection
+from .events import AssetKey
+from .run_request import PipelineRunReaction, RunRequest, SkipReason
+from .sensor_definition import (
+    DefaultSensorStatus,
+    MultiAssetSensorDefinition,
+    MultiAssetSensorEvaluationContext,
+    SensorDefinition,
+)
+from .target import ExecutableDefinition
+from .utils import check_valid_name
+
+if TYPE_CHECKING:
+    from dagster._core.definitions import AssetsDefinition, SourceAsset
+
+
+class UnresolvedAssetSensorDefinition:
+    def __init__(self):
+        pass
+
+    def resolve(
+        self,
+        assets: Sequence["AssetsDefinition"],
+        source_assets: Sequence["SourceAsset"],
+    ):
+        pass
+
+
+@experimental
+class UnresolvedAssetSensorDefinitionV1(UnresolvedAssetSensorDefinition):
+    def __init__(
+        self,
+        asset_selection: AssetSelection,
+        name: str,
+        minimum_interval_seconds: Optional[int] = None,
+        description: Optional[str] = None,
+        default_status: DefaultSensorStatus = DefaultSensorStatus.STOPPED,
+    ):
+        self._selection = asset_selection
+        self._name = name
+        self._minimum_interval_seconds = minimum_interval_seconds
+        self._description = description
+        self._default_status = default_status
+
+        super(UnresolvedAssetSensorDefinitionV1, self).__init__()
+
+    @property
+    def name(self):
+        return self._name
+
+    def _get_upstream_mapping(
+        self,
+        assets: Sequence["AssetsDefinition"],
+        source_assets: Sequence["SourceAsset"],
+    ):
+        """Computes a mapping of assets in self._selection to their parents in the asset graph"""
+        upstream = defaultdict(set)
+        selection_resolved = list(self._selection.resolve([*assets, *source_assets]))
+        for a in selection_resolved:
+            a_parents = list(AssetSelection.keys(a).upstream().resolve([*assets, *source_assets]))
+            # filter out a because upstream() includes the assets in the original AssetSelection
+            upstream[a] = {p for p in a_parents if p != a}
+        return upstream
+
+    def _get_parent_materializations(self, context, asset_key, parent_assets, cursor):
```

Comment:
I'm going to rename this to be more generic `_get_parent_updates` or something along those lines so that if/when we incorporate source asset versioning we can support source assets